File: python/src/pipelines/zebra_wings/models/classification.py
import json
import keras
import src.utils.dictionary_utils as du
import json
from random import shuffle
import numpy as np
import math
from keras.utils import to_categorical


class ClassificationModel:
    @staticmethod
    def setup(config, dataset_params):
        intents = dataset_params["intents"]
        max_words = dataset_params["maxWordsPerSentence"]
        embedding_dimensions = config["embeddingDimensions"]
        num_filters = config['numFilters']
        filter_sizes = config['filterSizes']
        drop = config["drop"]
        num_classes = len(intents)
        LEARNING_RATE = 0.0066  # use 1e-4 as default as alternative starting point
        ADAM_BETA_1 = 0.0025
        ADAM_BETA_2 = 0.1
        optimizer = keras.optimizers.Adam(
            lr=LEARNING_RATE, beta_1=ADAM_BETA_1, beta_2=ADAM_BETA_2)
        # Layer 1: Convolution + max pool
        inputs = keras.layers.Input(
            shape=(max_words, embedding_dimensions), name="embedded_words")
        convLayer1 = keras.layers.Conv1D(
            num_filters,
            filter_sizes[0],
            input_shape=(max_words, embedding_dimensions),
            kernel_initializer='random_normal',
            padding='valid',
            activation='relu',
        )(inputs)
        maxpool1 = keras.layers.MaxPooling1D(
            pool_size=max_words - filter_sizes[0] + 1,
            padding='valid'
        )(convLayer1)
        # Layer 2: Convolution + max pool
        convLayer2 = keras.layers.Conv1D(
            num_filters,
            filter_sizes[1],
            input_shape=(max_words, embedding_dimensions),
            kernel_initializer='random_normal',
            padding='valid',
            activation='relu',
        )(inputs)
        maxpool2 = keras.layers.MaxPooling1D(
            pool_size=max_words - filter_sizes[1] + 1,
            padding='valid'
        )(convLayer2)
        # Layer 3: Convolution + max pool
        convLayer3 = keras.layers.Conv1D(
            num_filters,
            filter_sizes[2],
            input_shape=(max_words, embedding_dimensions),
            kernel_initializer='random_normal',
            padding='valid',
            activation='relu',
        )(inputs)
        maxpool3 = keras.layers.MaxPooling1D(
            pool_size=max_words - filter_sizes[2] + 1,
            padding='valid'
        )(convLayer3)
        # Concatenation of all CNN layers on different levels and apply a fully connected dense layer
        concat = keras.layers.Concatenate(axis=1)(
            [maxpool1, maxpool2, maxpool3])
        flat = keras.layers.Flatten()(concat)
        dropout = keras.layers.Dropout(drop)(flat)
        flat_pool_1 = keras.layers.Flatten()(maxpool1)
        concat_for_classification = keras.layers.Concatenate(
            axis=1)([dropout, flat_pool_1])
        outputs = keras.layers.Dense(
            num_classes,
            activation='softmax',
        )(concat_for_classification)
        model = keras.models.Model(inputs=inputs, outputs=outputs)
        model.compile(
            optimizer=optimizer,
            loss='categorical_crossentropy',
            metrics=['accuracy'],
        )
        return model

    def __init__(self, config, dataset_params, embeddings_model, logger, pretrained_model=None):
        self.__config = config
        self.__dataset_params = dataset_params
        self.__embeddings_model = embeddings_model
        if (pretrained_model != None):
            self.__model = pretrained_model
        else:
            self.__model = ClassificationModel.setup(config, dataset_params)
        self.__logger = logger

    def keras_model(self):
        return self.__model

    def predict(self, sentences):
        prediction = []
        embedded_sentences = self.__embeddings_model.embed(sentences)
        output = self.__model.predict(embedded_sentences)
        intents = self.__dataset_params['intents']
        for sidx, s in enumerate(output):
            max_intent_index = s.argmax()
            prediction.append({
                'intent': intents[max_intent_index], 'confidence': s[max_intent_index], 'sentence': sentences[sidx],
            })
        return prediction

    def train(self, train_dataset):
        train_data = du.chunks(
            train_dataset['trainX'], self.__config['batchSize'], train_dataset['trainY'])
        n_batches = math.ceil(
            len(train_dataset['trainX'])/self.__config['batchSize'])
        self.__logger(f'Start training classification model!')
        enough_accuracy_reached = False
        m = self.__model
        for idx, t_chunk in enumerate(train_data):
            if enough_accuracy_reached:
                break
            x = self.__embeddings_model.embed(t_chunk[0])
            y = to_categorical(np.array(t_chunk[1], dtype=np.int32))
            self.__logger(f'Training batch {idx+1}.')
            m.fit(
                x=x,
                y=y,
                batch_size=self.__config['batchSize'],
                shuffle=True,
                epochs=self.__config['epochs'],
                verbose=0,
                validation_split=self.__config['trainingValidationSplit'],
            )
            self.__logger(
                f'Trained {m.history.epoch[-1]+1} epochs on batch {idx + 1} of {n_batches}')
            self.__logger(
                f'Training Loss: {m.history.history["loss"][-1]} | Training Accuracy: {m.history.history["acc"][-1]}')
            self.__logger(
                f'Validation Loss: {m.history.history["val_loss"][-1]} | Validation Accuracy: {m.history.history["val_acc"][-1]}')
            self.__logger(
                '==================================================================================================')
            if (self.__config["lossThresholdToStopTraining"] > 0 and
                m.history.history["loss"][-1] < self.__config["lossThresholdToStopTraining"] and
                    m.history.history["val_loss"][-1] < self.__config["lossThresholdToStopTraining"]):
                enough_accuracy_reached = True
                self.__logger(
                    f'Enough accuracy reached! Ending training after batch {idx + 1} of {n_batches}')
                self.__logger(
                    '==================================================================================================')

    def test(self, test_examples, results_handler=None):
        chunks = du.chunks(
            test_examples['testX'], 100, test_examples['testY'])
        handler = results_handler if results_handler != None else self.__default_results_logger
        stats = {'correct': 0, 'wrong': 0, 'lowConfidence': 0}
        for t_chunk in chunks:
            x = t_chunk[0]  # sentences
            y = t_chunk[1]  # intents code per sentence
            predictions = self.predict(x)
            handler(x, y, predictions, stats)
        return stats

    def __default_results_logger(self, x, y, o, stats):
        for i, s in enumerate(x):
            intent = self.__dataset_params['intents'][y[i]]
            correct = o[i]['intent'] == intent
            if (o[i]['confidence'] < self.__config['lowConfidenceThreshold']):
                stats['lowConfidence'] += 1
                self.__logger(
                    f'LOW CONFIDENCE (intent: {o[i]["intent"]}, confidence: {o[i]["confidence"]}) - {s}')
            elif (correct):
                stats['correct'] += 1
                # self.__logger(
                #     f'CORRECT (intent: {o[i]['intent']}, confidence: {o[i]['confidence']}) - {s}')
            else:
                stats['wrong'] += 1
                self.__logger(
                    f'WRONG (intent: {o[i]["intent"]}, confidence: {o[i]["confidence"]}) - {s}')
        return stats

File: typescript/src/pipelines/zebraWings/models/classification.ts
import * as tf from '@tensorflow/tfjs';
import { chunk } from 'lodash';
import * as types from '../../../types';
import { EmbeddingsModel } from '../embeddings/EmbeddingsModel';

export default class ClassificationModel extends types.PipelineModel implements types.IPipelineModel {
    private static setup(
        config: types.IClassificationModelParams & types.IDefaultModelParams,
        { maxWordsPerSentence: maxWords, intents }: types.IDatasetParams
    ) {
        const numClasses = intents.length;
        const LEARNING_RATE = 0.0066; // use 1e-4 as default as alternative starting point
        const ADAM_BETA_1 = 0.0025;
        const ADAM_BETA_2 = 0.1;
        const optimizer = tf.train.adam(LEARNING_RATE, ADAM_BETA_1, ADAM_BETA_2);
        // Layer 1: Convolution + max pool
        const input = tf.input({ dtype: 'float32', shape: [maxWords, config.embeddingDimensions] });
        const convLayer1 = tf.layers
            .conv1d({
                activation: 'relu',
                filters: config.numFilters,
                inputShape: [maxWords, config.embeddingDimensions],
                kernelInitializer: 'randomNormal',
                kernelSize: [config.filterSizes[0]],
                padding: 'valid'
            })
            .apply(input);
        const maxpool1 = tf.layers
            .maxPooling1d({
                padding: 'valid',
                poolSize: maxWords - config.filterSizes[0] + 1
            })
            .apply(convLayer1) as tf.SymbolicTensor;
        // Layer 2: Convolution + max pool
        const convLayer2 = tf.layers
            .conv1d({
                activation: 'relu',
                filters: config.numFilters,
                inputShape: [maxWords, config.embeddingDimensions],
                kernelInitializer: 'randomNormal',
                kernelSize: [config.filterSizes[1]],
                padding: 'valid'
            })
            .apply(input);
        const maxpool2 = tf.layers
            .maxPooling1d({
                padding: 'valid',
                poolSize: maxWords - config.filterSizes[1] + 1
            })
            .apply(convLayer2) as tf.SymbolicTensor;
        // Layer 3: Convolution + max pool
        const convLayer3 = tf.layers
            .conv1d({
                activation: 'relu',
                filters: config.numFilters,
                inputShape: [maxWords, config.embeddingDimensions],
                kernelInitializer: 'randomNormal',
                kernelSize: [config.filterSizes[2]],
                padding: 'valid'
            })
            .apply(input);
        const maxpool3 = tf.layers
            .maxPooling1d({
                padding: 'valid',
                poolSize: maxWords - config.filterSizes[2] + 1
            })
            .apply(convLayer3) as tf.SymbolicTensor;
        // Concatenation of all CNN layers on different levels and apply a fully connected dense layer
        const concatLayer = tf.layers.concatenate({ axis: 1 }).apply([maxpool1, maxpool2, maxpool3]);
        const flat = tf.layers.flatten().apply(concatLayer);
        const dropOut = tf.layers.dropout({ rate: config.drop }).apply(flat) as tf.SymbolicTensor;
        const flatPool1 = tf.layers.flatten().apply(maxpool1) as tf.SymbolicTensor;
        const concatForClassification = tf.layers.concatenate({ axis: 1 }).apply([dropOut, flatPool1]);
        const outClassification = tf.layers
            .dense({
                activation: 'softmax',
                units: numClasses
            })
            .apply(concatForClassification) as tf.SymbolicTensor;
        const model = tf.model({ inputs: input, outputs: outClassification });
        model.compile({
            loss: 'categoricalCrossentropy',
            metrics: ['accuracy'],
            optimizer
        });
        return model;
    }

    private config: types.IClassificationModelParams & types.IDefaultModelParams;
    private datasetParams: types.IDatasetParams;
    private model: tf.Model;
    private embeddingsModel: EmbeddingsModel;
    private logger: types.IPipelineModelLogger;
    private classificationTrainStatsHandler: types.ITrainStatsHandler['classification'] | undefined;
    constructor(
        config: types.IClassificationModelParams & types.IDefaultModelParams,
        datasetParams: types.IDatasetParams,
        embeddingsModel: EmbeddingsModel,
        logger: types.IPipelineModelLogger,
        pretrainedModel?: tf.Model,
        classificationTrainStatsHandler?: types.ITrainStatsHandler['classification']
    ) {
        super();
        this.config = config;
        this.datasetParams = datasetParams;
        this.embeddingsModel = embeddingsModel;
        this.model = pretrainedModel ? pretrainedModel : ClassificationModel.setup(this.config, this.datasetParams);
        this.logger = logger;
        this.classificationTrainStatsHandler = classificationTrainStatsHandler;
    }

    public tfModel = () => this.model;

    public predict = (sentences: string[]): types.IClassificationPred[] => {
        const prediction = [] as types.IClassificationPred[];
        tf.tidy(() => {
            const embeddedSentences = this.embeddingsModel.embed(sentences);
            const output = this.model.predict(embeddedSentences) as tf.Tensor<tf.Rank>;
            const d = output.dataSync() as Float32Array;
            output.dispose();
            embeddedSentences.dispose();
            const intents = this.datasetParams.intents;
            sentences.forEach((s, i) => {
                const preds = d.slice(i * intents.length, i * intents.length + intents.length);
                const sentencePreds: types.IClassificationPred[] = [];
                preds.forEach((p, idx) =>
                    sentencePreds.push({
                        confidence: p,
                        intent: intents[idx],
                        sentence: s
                    })
                );
                sentencePreds.sort((a: any, b: any) => (a.confidence > b.confidence ? -1 : 1));
                prediction.push(sentencePreds[0]);
            });
        });
        return prediction;
    };

    public train = async (trainDataset: types.ITrainingParams): Promise<void> => {
        const trainYChunks = chunk(trainDataset.trainY, this.config.batchSize);
        const trainXChunks = chunk(trainDataset.trainX, this.config.batchSize);
        this.logger.log('Start training classification model!');
        const m = this.model;
        let enoughAccuracyReached = false;
        for (const [index, xChunk] of trainXChunks.entries()) {
            if (enoughAccuracyReached) {
                return;
            }
            const embeddedSentences = this.embeddingsModel.embed(xChunk);
            const dataLabels = tf.tensor1d(trainYChunks[index], 'int32');
            const hotEncodedLabels = tf.oneHot(dataLabels, this.datasetParams.intents.length);
            await m.fit(embeddedSentences, hotEncodedLabels, {
                // batchSize: this.config.batchSize,
                callbacks: { onBatchEnd: tf.nextFrame },
                epochs: this.config.epochs,
                shuffle: true,
                validationSplit: this.config.trainingValidationSplit
            });
            dataLabels.dispose();
            embeddedSentences.dispose();
            hotEncodedLabels.dispose();
            const h = m.history.history;
            const c = h.val_loss.length - 1;
            if (this.classificationTrainStatsHandler) {
                this.classificationTrainStatsHandler({
                    batch: index + 1,
                    batchEpochs: m.history.epoch.length,
                    currentBatchSize: trainXChunks[index].length,
                    tensorsInMemory: tf.memory().numTensors,
                    totalBatches: trainXChunks.length,
                    trainingAccuracy: h.acc[c],
                    trainingLoss: h.loss[c],
                    validationAccuracy: h.val_acc[c],
                    validationLoss: h.val_loss[c]
                });
            }
            this.logger.log(`Trained ${m.history.epoch.length} epochs on batch ${index + 1} of ${trainXChunks.length}`);
            this.logger.log(`Training Loss: ${h.loss[c]} | Training Accuracy: ${h.acc[c]}`);
            this.logger.log(`Validation Loss: ${h.val_loss[c]} | Validation Accuracy: ${h.val_acc[c]}`);
            this.logger.warn(`(Memory) Number of tensors in memory at the end of batch: ${tf.memory().numTensors}`);
            this.logger.log('==================================================================================================');
            if (
                this.config.lossThresholdToStopTraining &&
                h.loss[c] < this.config.lossThresholdToStopTraining &&
                h.val_loss[c] < this.config.lossThresholdToStopTraining
            ) {
                enoughAccuracyReached = true;
                this.logger.warn(`Enough accuracy reached! Ending training after batch ${index + 1} of ${trainXChunks.length}`);
                this.logger.log('==================================================================================================');
            }
        }
    };

    public test = async (
        testExamples: types.ITestingParams,
        resultsHandler?: types.ITestPredictionsHandler
    ): Promise<types.IPredictionStats> => {
        const handler = resultsHandler ? resultsHandler : this.defaultResultsLogger;
        const stats: types.IPredictionStats = { correct: 0, wrong: 0, lowConfidence: 0 };
        const x = chunk(testExamples.testX, this.config.batchSize);
        const y = chunk(testExamples.testY, this.config.batchSize);
        for (const [i, sentences] of x.entries()) {
            const predictions = this.predict(sentences);
            handler(sentences, y[i], predictions, stats);
        }
        return stats;
    };

    private defaultResultsLogger = (
        x: types.ITestingParams['testX'],
        y: types.ITestingParams['testY'],
        o: types.IClassificationPred[],
        stats: types.IPredictionStats
    ): types.IPredictionStats => {
        x.forEach((s, i) => {
            const intent = this.datasetParams.intents[y[i]];
            const correct = o[i].intent === intent;
            if (o[i].confidence < this.config.lowConfidenceThreshold) {
                if (stats.lowConfidence === undefined) {
                    return;
                }
                stats.lowConfidence++;
                this.logger.warn(`LOW CONFIDENCE (intent: ${o[i].intent}, confidence: ${o[i].confidence}) - ${s}`);
            } else if (correct) {
                stats.correct++;
                this.logger.debug(`CORRECT (intent: ${o[i].intent}, confidence: ${o[i].confidence}) - ${s}`);
            } else {
                stats.wrong++;
                this.logger.error(`WRONG (intent: ${o[i].intent}, confidence: ${o[i].confidence}) - ${s}`);
            }
        });
        return stats;
    };
}
